python_modules/dagster/dagster/_utils/calculate_data_time.py

```diff
@@ -0,0 +1,107 @@
+        # no record available, so the data time for all keys upstream of here is None
+        if record is None:
+            return {k: (None, None) for k in required_keys}
+
+        cur_tags = instance.get_asset_event_tags(record.storage_id)
+
+        cur_key = cast(AssetKey, record.asset_key)
+        cur_key_str = cur_key.to_user_string()
+
+        # grab the existing upstream data times already calculated for this record (if any)
+        if UPSTREAM_MATERIALIZATION_TIMES_TAG in cur_tags:
+            known_times = cur_tags[UPSTREAM_MATERIALIZATION_TIMES_TAG]
+        else:
+            known_times = {}
+
+        # this is a required key to track, so grab its record
+        if cur_key in required_keys:
+            known_times[cur_key_str] = (
+                record.storage_id,
+                record.event_log_entry.timestamp,
+            )
+
+        # not all required data times are present in the tags
+        if required_keys - set(known_times.keys()):
+
+            # find all the required keys that are upstream of this key
+            required_upstream_keys = set()
+            for upstream_key_str in _upstream_key_str_iterator(cur_key_str):
+                upstream_key = AssetKey.from_user_string(upstream_key_str)
+                if upstream_key in required_keys:
+                    required_upstream_keys.add(upstream_key)
+
+            # find the upstream times of each of the parents of this asset
+            for parent_key_str in upstream_asset_key_mapping[cur_key_str]:
+                parent_key = AssetKey.from_user_string(parent_key_str)
+
+                # get the most recent asset materialization for this parent which happened before
+                # the current record
+                parent_records = instance.get_event_records(
+                    EventRecordsFilter(
+                        event_type=DagsterEventType.ASSET_MATERIALIZATION,
+                        asset_key=parent_key,
+                        before_cursor=record.storage_id,
+                    ),
+                    ascending=False,
+                    limit=1,
+                )
+                latest_parent_record = parent_records[0] if parent_records else None
+
+                # recurse to find the data times of this parent
+                for key, tup in _get_upstream_times_and_ids(
+                    latest_parent_record, required_upstream_keys
+                ).items():
+                    tup = tuple(tup)
+                    # if root data is missing, override other values
+                    if tup == (None, None) or known_times.get(key) == (None, None):
+                        known_times[key] = (None, None)
+                    else:
+                        known_times[key] = min(tuple(known_times.get(key, tup)), tup)
+
+        # cache asset event in the database
+        instance.add_asset_event_tags(
+            record.storage_id, {UPSTREAM_MATERIALIZATION_TIMES_TAG: known_times}
```

Comment:
Have you considered having a tag per upstream asset? That would allow us to do queries like "what materializations are downstream of this materialization".  I know we wouldn't be able to do that immediately, because it won't always be cached, but I suspect there's a road to it.